### PR TITLE
Stack Unwinding macOS: implemented -generate_unwind

### PR DESCRIPTION
**Pull request by avniculae**

### CODE REVIEW

File: macOS/unwindmacos.cpp [hunk 1]
```diff
@@ -20,49 +20,116 @@ limitations under the License.
...
+      last_lookup.SetEncoding(0, prev(it)->first, -1);
+    } else {
+      last_lookup.SetEncoding(prev(it)->second, prev(it)->first, it->first - 1);
+    }
+  }
+}
+
+void UnwindDataMacOS::LookupLSDA(size_t original_address) {
+  if (last_lookup.IsLsdaMiss(original_address)) {
+    if (!lsda_map.empty() && last_lookup.encoding & UNWIND_HAS_LSDA) {
```
Comment (j00ru) [inline macOS/unwindmacos.cpp:49]: A condition of `x && y & z` is a bit hard to read even if you know the evaluation order of the operators. I'd put `last_lookup.encoding & UNWIND_HAS_LSDA` in parenthesis here.

File: macOS/unwindmacos.cpp [hunk 2]
```diff
@@ -20,49 +20,116 @@ limitations under the License.
...
+  }
+}
+
+void UnwindGeneratorMacOS::SanityCheckUnwindHeader(ModuleInfo *module) {
+  UnwindDataMacOS *unwind_data = (UnwindDataMacOS *)module->unwind_data;
+  unwind_info_section_header *unwind_section_header = unwind_data->unwind_section_header;
+
+  if (unwind_section_header->version != UNWIND_SECTION_VERSION) {
+    FATAL("Unexpected version (%u) in Unwind Section Header",
+          unwind_data->unwind_section_header->version);
```
Comment (j00ru) [inline macOS/unwindmacos.cpp:114]: You can skip the `unwind_data->` here as you already have a pointer to it in `unwind_section_header`.

File: macOS/unwindmacos.cpp [hunk 3]
```diff
@@ -121,102 +187,177 @@ void UnwindGeneratorMacOS::OnModuleUninstrumented(ModuleInfo *module) {
...
+size_t UnwindGeneratorMacOS::WriteCIE(ModuleInfo *module,
+                                      const char *augmentation,
+                                      size_t personality_addr) {
+  personality_addr = GetCustomPersonality(module, personality_addr);
+  
+  ByteStream cie;
+  cie.PutBytes<uint32_t>(0);          // CIE id, must be zero
+  cie.PutBytes<uint8_t>(3);           // version
+  cie.PutString(augmentation);        // augmentation string
+  cie.PutEncodedULEB128Bytes(1);      // code aligment factor, ULEB128 encoded
```
Comment (j00ru) [inline macOS/unwindmacos.cpp:206]: s/aligment/alignment

here and below

File: macOS/unwindmacos.cpp [hunk 4]
```diff
@@ -288,64 +470,110 @@ bool UnwindGeneratorMacOS::HandleBreakpoint(ModuleInfo* module, void *address) {
...
+
+    if ((saved_registers_locations & 0x7) != UNWIND_X86_64_REG_NONE) {
+      fde->PutEncodedULEB128Bytes(saved_registers_cfa_offset);
+    }
+
+    saved_registers_cfa_offset += 8;
+    saved_registers_locations >>= 3;
+  }
+
+//  fde->PutBytes<uint8_t>(DW_CFA_def_cfa);
```
Comment (j00ru) [inline macOS/unwindmacos.cpp:519]: Why is this commented out? I'd either remove it completely or add a comment to explain.

File: macOS/unwindmacos.cpp [hunk 5]
```diff
@@ -288,64 +470,110 @@ bool UnwindGeneratorMacOS::HandleBreakpoint(ModuleInfo* module, void *address) {
...
+}
+
+void UnwindGeneratorMacOS::WriteDWARFInstructions(ByteStream *fde, uint32_t encoding) {
+  uint32_t mode = encoding & UNWIND_X86_64_MODE_MASK;
+  switch(mode) {
+    case UNWIND_X86_64_MODE_RBP_FRAME:
+      WriteDWARFInstructionsRBP(fde, encoding);
+      break;
+    default:
+      WARN("Unsupported encoding");
```
Comment (j00ru) [inline macOS/unwindmacos.cpp:531]: Perhaps could be useful to print out the value of the unsupported encoding here?

File: macOS/unwindmacos.cpp [hunk 6]
```diff
@@ -414,8 +619,104 @@ size_t UnwindGeneratorMacOS::MaybeRedirectExecution(ModuleInfo* module, size_t I
...
+  size_t code_size_before = module->instrumented_code_allocated;
+
+  size_t new_personality_addr = tinyinst_.GetCurrentInstrumentedAddress(module);
+  
+  unsigned char assembly_part1[] = {
+    // function prologue
+    0x55, // push rbp
+    0x48, 0x89, 0xE5, // mov rbp, rsp
+    // save registers we're modifying
+    0x53,
```
Comment (j00ru) [inline macOS/unwindmacos.cpp:646]: These four instructions are missing mnemonics in comments? :)

File: macOS/unwindmacos.cpp [hunk 7]
```diff
@@ -414,8 +619,104 @@ size_t UnwindGeneratorMacOS::MaybeRedirectExecution(ModuleInfo* module, size_t I
...
+    0x48, 0x85, 0xDB, // test   rbx,rbx
+    0x0F, 0x84, 0x07, 0x00, 0x00, 0x00, // je no_original_personality
+    // call original personality function
+    0xFF, 0xD3, // call rbx
+    0xE9, 0x07, 0x00, 0x00, 0x00, // jmp end
+    // no_original_personality:
+    0x48, 0xC7, 0xC0, 0x08, 0x00, 0x00, 0x00, // mov rax,8 (_URC_CONTINUE_UNWIND )
+    // end:
+    //restore registers
+    0x41, 0x5E,
```
Comment (j00ru) [inline macOS/unwindmacos.cpp:703]: Some missing comments with mnemonics here as well.

File: macOS/debugger.cpp [hunk 8]
```diff
@@ -1002,30 +1004,45 @@ void *Debugger::GetSymbolAddress(void *base_address, char *symbol_name) {
...
+  for (int i = 0; i < symtab_cmd->nsyms && !symbol_address; ++i) {
+    nlist_64 curr_symbol = *(nlist_64*)curr_symbol_address;
+    if ((curr_symbol.n_type & N_TYPE) == N_SECT) {
+      char *curr_sym_name = NULL;
+      if (!in_shared_cache) {
+        curr_sym_name = strtab + curr_symbol.n_un.n_strx;
+      } else {
+        std::string curr_sym_name_string;
+        mach_target->ReadCString(strtab_addr + curr_symbol.n_un.n_strx, curr_sym_name_string);
+        curr_sym_name = (char*)curr_sym_name_string.c_str();
```
[the remark this answers is not in the document]
Comment (avniculae) [reply macOS/debugger.cpp:1032]: This was already solved in the latest commit. Thanks for spotting it and sorry! :)

File: macOS/unwindmacos.cpp (hunk 6 above)
Comment (ifratric) [reply macOS/unwindmacos.cpp:646]: This is entirely my fault as I wrote this assembly snippets. Anyway, the instructions here are 
```
53                      push   rbx
41 54                   push   r12
41 55                   push   r13
41 56                   push   r14
```

File: macOS/unwindmacos.cpp (hunk 7 above)
Comment (ifratric) [reply macOS/unwindmacos.cpp:703]:
```
41 5e                   pop    r14
41 5d                   pop    r13
41 5c                   pop    r12
5b                      pop    rbx
```

**Comment by ifratric** [comment]
Thanks, merging!

**Pull request merged**